### PR TITLE
Ctrl iface zephyr

### PR DESCRIPTION
**Pull request by krish2718**
Use Eloop and sockets for WPA CLI
  
  In Linux WPA CLI is a standalone application that uses sockets to
  communicate with ctrl_iface (uses Eloop), in Zephyr we have replaced
  sockets with direct function calls, but this introduces concurrency as
  we don't go through Eloop.
  
  Implement Zephyr's version of WPA CLI and control interface using
  socketpair UNIX style sockets, here WPA CLI can be invoked as a standard
  function rather than an executable.
  
  Also, as WPA CLI is meant for internal usage, remove exposing of limited
  set of commands and instead make it a pass through to accept any
  command.
  
  Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

### CODE REVIEW

File: wpa_supplicant/wpa_cli_zephyr.c
Comment (maxd-nordic) [inline wpa_supplicant/wpa_cli_zephyr.c:1]: This huge file should not be copied from `wpa_cli.c` - better find a way to make the original work.

Comment (krish2718) [reply wpa_supplicant/wpa_cli_zephyr.c:1]: We cannot re-use the file as it was intended to be a separate application with interactive usage, we could split the wpa_cli commands in to a common file and then re-use but that would be make maintaining it difficult which is why I forked it.

Comment (krish2718) [reply wpa_supplicant/wpa_cli_zephyr.c:1]: For now I have not split the common code in to a separate file, to keep upmerging simple.

**Review by sr1dh48r — APPROVED** [review]

File: wpa_supplicant/ctrl_iface.c
Comment (wbober) [inline wpa_supplicant/ctrl_iface.c:11487]: What's wrong with that `memset`?

**Comment by wbober** [comment]
Good job! If my understanding is correct you're also reverting a number of changes that were introduced when porting to Zephyr but now aren't needed, hence the code will be more aligned with vanilla wpa supplicant.

Two comments to the structure for your consideration:
1) I think it would really help do define API naming convention for functions that are zephyr specific, public, and safe for any zephyr thread to call. There is a mixture of naming prefixes that, at least I, find hard to comprehend. 

You could, for example, use `z_wpa_*` prefix. You could then rename:

```
zephyr_supp_status -> z_wpa_supplicant_status
zephyr_supp_disconnect -> z_wpa_supplicant_disconnect
zephyr_supp_connect -> z_wpa_supplicant_connect
```

2) I would keep the new zephyr related code in `zephyr/src` for the time being. Easier to understand what is added as part of the zephyr port.

File: wpa_supplicant/wpa_cli_zephyr.h
Comment (wbober) [inline wpa_supplicant/wpa_cli_zephyr.h:23]: I would rename to `z_wpa_ctrl_init()`, `z_wpa_ctrl_deinit()` and `z_wpa_cli()`.

File: zephyr/src/supp_api.c
Comment (wbober) [inline zephyr/src/supp_api.c:197]: Should this be part of the `wpa_cli_zephyr.h` API as, for example, `z_wpa_cli_cmd_v`?

Comment (maxd-nordic) [reply zephyr/src/supp_api.c:197]: I would appreciate something like this to be exposed for ease of use. :)

File: zephyr/src/supp_api.c
Comment (maxd-nordic) [inline zephyr/src/supp_api.c:314]: Don't forget to set key_mgmt: `call_wpa_cli("set_network %d key_mgmt SAE", resp.network_id);`

File: wpa_supplicant/wpa_cli_zephyr.h
Comment (maxd-nordic) [inline wpa_supplicant/wpa_cli_zephyr.h:14]: These internal includes are problematic. They should not be included in a header the app might try to use.

File: zephyr/src/supp_api.c
Comment (maxd-nordic) [inline zephyr/src/supp_api.c:201]: unused now

File: zephyr/src/supp_api.c
Comment (maxd-nordic) [inline zephyr/src/supp_api.c:292]: unused now

File: zephyr/src/supp_api.c
Comment (maxd-nordic) [inline zephyr/src/supp_api.c:287]: unused now

File: zephyr/src/supp_api.c
Comment (maxd-nordic) [inline zephyr/src/supp_api.c:261]: send_wpa_supplicant_dummy_event is completely unused now

File: wpa_supplicant/ctrl_iface_zephyr.c
Comment (maxd-nordic) [inline wpa_supplicant/ctrl_iface_zephyr.c:15]: unused now

File: wpa_supplicant/wpa_cli_zephyr.c
Comment (maxd-nordic) [inline wpa_supplicant/wpa_cli_zephyr.c:135]: unused

**Comment by glarsennordic** [comment]
Chiming in, I'm reviewing a related PR,

https://github.com/nrfconnect/sdk-nrf/pull/9261

Right now that PR is forced to make heavy use of `call_wpa_cli`, but that usage could (in my opinion) be reduced significantly if `zephyr_supp_connect` were broken down into smaller sub-functions, related to each step of setting up and then associating with a network.

I would like to suggest that `zephyr_supp_connect` be broken down in that way.

Perhaps there should even be a callable function for each individual command the CLI accepts?

File: zephyr/src/supp_api.c
Comment (glarsennordic) [inline zephyr/src/supp_api.c:81]: AFAICT, quote will never have any value other than 0 in this function, is that intended?

Comment (krish2718) [reply zephyr/src/supp_api.c:81]: Yes, this can be removed.

**Comment by krish2718** [comment]
> Chiming in, I'm reviewing a related PR,
> 
> [nrfconnect/sdk-nrf#9261](https://github.com/nrfconnect/sdk-nrf/pull/9261)
> 
> Right now that PR is forced to make heavy use of `call_wpa_cli`, but that usage could (in my opinion) be reduced significantly if `zephyr_supp_connect` were broken down into smaller sub-functions, related to each step of setting up and then associating with a network.
I am not sure I understand how breaking into smaller functions reduces `call_wpa_cli`, since we need to call `call_wpa_cli` for each of them anyway, with this new interface WPA supplicant only offers individual settings.
> 
> I would like to suggest that `zephyr_supp_connect` be broken down in that way.
> 
> Perhaps there should even be a callable function for each individual command the CLI accepts?
Well, sure, but we are trying to aggregate and expose a smaller set to the applications, we already have `wpa_cli` (disabled by default and can be enabled by `CONFIG_WPA_CLI=y` that exposes full commands of WPA supplicant.)
> 
> See this comment: [nrfconnect/sdk-nrf#9261 (comment)](https://github.com/nrfconnect/sdk-nrf/pull/9261#discussion_r1080712094)

File: wpa_supplicant/wpa_cli_zephyr.c
Comment (krish2718) [reply wpa_supplicant/wpa_cli_zephyr.c:3882]: `%*` is used to ignore the values

Comment (krish2718) [reply wpa_supplicant/wpa_cli_zephyr.c:3882]: had to escape % `%%*`

**Comment by krish2718** [comment]
> These would still be higher-level than direct wpa_cli commands. I think since it is such a common design pattern to add a network before actually connecting, exposing those two as separate steps (and also including a way to remove/deselect networks) could make sense

Sure, but the APIs in `zephyr/src/supp_api.c` directly map to the Zephyr's wifi_mgmt offload APIs, we can divide the APIs in case you want it to used for other usecases, but for now wifi_mgmt is the only case.

File: wpa_supplicant/wpa_cli_zephyr.c
Comment (krish2718) [reply wpa_supplicant/wpa_cli_zephyr.c:1]: I have bitten the bullet and did this.

File: zephyr/src/supp_api.c
Comment (krish2718) [reply zephyr/src/supp_api.c]: As its a copy, I left the code unchanged.

File: zephyr/src/supp_api.c
Comment (maxd-nordic) [inline zephyr/src/supp_api.c:119]: is there a corresponding change in sdk-zephyr?

Comment (krish2718) [reply zephyr/src/supp_api.c:119]: Yes, the `sdk-nrf` PR linked has the manifest change.

Comment (maxd-nordic) [reply zephyr/src/supp_api.c:119]: :+1:

**Comment by glarsennordic** [comment]
> Sure, but the APIs in zephyr/src/supp_api.c directly map to the Zephyr's wifi_mgmt offload APIs, we can divide the APIs in case you want it to used for other usecases, but for now wifi_mgmt is the only case.

I would be inclined to do so, but I also recognize that's my personal opinion. I'll leave the final decision up to you and @maxd-nordic and the other reviewers

**Comment by krish2718** [comment]
To fix CI where one of the Matter sample was failing, disabled all non-zephyr commands by default using CONFIG_WPA_CLI, this saves about 2% flash.

**Review by wbober — APPROVED** [review]

**Review by glarsennordic — APPROVED** [review]

File: wpa_supplicant/ctrl_iface_zephyr.c
Comment (krish2718) [reply wpa_supplicant/ctrl_iface_zephyr.c:43]: The original code has this on the stack, but to reduce stack sizes moved this to HEAP.

File: wpa_supplicant/ctrl_iface_zephyr.c
Comment (krish2718) [reply wpa_supplicant/ctrl_iface_zephyr.c:133]: We need to implement these dummy because they will be called from the generic control interface code, but for now these are no-ops.

File: wpa_supplicant/wpa_cli.c
Comment (krish2718) [reply wpa_supplicant/wpa_cli.c:133]: Yes, keeping changes minimal for easier upmerge.

File: wpa_supplicant/wpa_cli_cmds.c
Comment (krish2718) [reply wpa_supplicant/wpa_cli_cmds.c:67]: Yes, but we don't use `wpa_cli.c` in Zephyr.

File: wpa_supplicant/wpa_cli_zephyr.c
Comment (krish2718) [reply wpa_supplicant/wpa_cli_zephyr.c:114]: Again, as this is a copy of the Zephyr's code, avoiding local changes.

File: wpa_supplicant/ctrl_iface_zephyr.c
Comment (krish2718) [reply wpa_supplicant/ctrl_iface_zephyr.c:11]: These are internal functions, so, do we need to add docstrings?

File: wpa_supplicant/wpa_cli.c
Comment (krish2718) [reply wpa_supplicant/wpa_cli.c:66]: No, we don't use `wpa_cli.c` in Zephyr.

File: wpa_supplicant/wpa_cli_cmds.c
Comment (krish2718) [reply wpa_supplicant/wpa_cli_cmds.c:3680]: Yes, this is to add space b/w all possible combinations, its from the upstream code.

**Review by sachinthegreen — APPROVED** [review]

File: wpa_supplicant/ctrl_iface_zephyr.c
Comment (glarsennordic) [reply wpa_supplicant/ctrl_iface_zephyr.c:11]: You're right, docstrings are probably not required, but some comments at least would be a good idea, just to help any prospective readers out

I won't block this PR on that, though

File: wpa_supplicant/ctrl_iface_zephyr.c
Comment (glarsennordic) [reply wpa_supplicant/ctrl_iface_zephyr.c:133]: I would recommend some comments indicating that

File: wpa_supplicant/ctrl_iface_zephyr.c
Comment (krish2718) [reply wpa_supplicant/ctrl_iface_zephyr.c:11]: > but some comments at least would be a good idea, just to help any prospective readers out

Helpful comments are strictly against OSS philosophy :)

File: wpa_supplicant/wpa_cli_zephyr.c
Comment (glarsennordic) [reply wpa_supplicant/wpa_cli_zephyr.c:115]: Where is it copied from in Zephyr, and could it be included instead?

And is the always-zero quote also an attribute of this copy?

Comment (krish2718) [reply wpa_supplicant/wpa_cli_zephyr.c:115]: https://github.com/nrfconnect/sdk-hostap/pull/32/files#diff-226f1a66d69db480c0d31c83165ca0adb98ec2c93cbe13f7abdc958b0ddd29f2R106 this comment points to Zephyr's original.

Now that I see, I had removed stripping off the escapes, you are right, this had a local change, will raise a PR to remove quote completely.

**Pull request merged**
